Adding standard, xo support; improving linter detection

chrisdothtml opened the pull request:
Fixes #13. Adding `standard` support. ~for #13~

Also, added support for detecting dependencies/devDependencies since a lot of users of `standard` won't have any configuration (kinda the whole point of standard, lol), so detecting based on whether they have any config in their `package.json` isn't super reliable. With this addition, I also added a couple more checks in the `eslint` detector

chrisdothtml committed:
modifying tryPackageJson to check both config and deps

chrisdothtml committed:
adding support for detecting standard

chrisdothtml committed:
adding more props to check for in eslint detector

chrisdothtml committed:
modifying detectors to pass array of props to tryPackageJson

chrisdothtml committed:
adding standard to package keywords

chrisdothtml committed:
fixing lint errors

chrisdothtml committed:
modifying eslint test to get coverage back to 100%

chrisdothtml commented:
~~Looks like CI is failing because of gotwarlost/istanbul#841. Not sure what the best workaround is other than using their alpha :/~~

~~EDIT:~~
~~Looks like that package is no longer maintained, and [istanbuljs/nyc](https://github.com/istanbuljs/nyc) is now istanbul's cli~~

chrisdothtml committed:
adding standard to readme list

chrisdothtml committed:
removing object destructuring

chrisdothtml commented:
Ignore my previous comment. It was my object destructuring that was breaking CI in node v4

chrisdothtml commented:
@satazor any feedback on this?

satazor commented:
I couldn't still review this MR but I will as soon as I got some free time. I'm sorry about that.

chrisdothtml commented:
ping :)

It's a pretty straightforward PR; the diff makes it look bigger than it is. The main change is in [tryPackageJson.js](https://github.com/IndigoUnited/node-detect-repo-linters/pull/25/files#diff-6b1896cd1fa5f79391f26e2ab373bd99)

chrisdothtml commented:
@satazor The reason I made this PR is because it impacts the quality score of packages on npm (and therefore their placement in the search). Packages that use StandardJS are currently being penalized because this package isn't able to detect them.

I would really appreciate it if you could provide feedback on this PR soon

satazor commented on lib/eslint.js:
```diff
@@ -10,7 +10,7 @@ function detectEslint(dir) {
     return Promise.all([
         tryFiles(paths),
-        tryPackageJson(dir, 'eslintConfig'),
+        tryPackageJson(dir, ['eslint', 'eslintConfig', 'eslintIgnore']),
```
I couldn't find in the docs mentioning the `eslint` prop: https://eslint.org/docs/user-guide/configuring
Could you point to a reference?

satazor replied:
Oh I see, you want to check if there's a `eslint` dependency. See my comment below.

satazor replied:
Perhaps we should analyze all other linters and do the same for every linter that does not require a `config` prop in package.json and just the dependency installed?

chrisdothtml replied:
> Perhaps we should analyze all other linters and do the same for every linter that does not require a `config` prop in package.json and just the dependency installed?

Agreed. I only updated eslint because I was already familiar with it

satazor commented on lib/util/tryPackageJson.js:
```diff
@@ -3,9 +3,22 @@
+function tryPackageJson(dir, props) {
     return loadJsonFile(path.join(dir, 'package.json'))
-    .then((json) => !!json[prop], (err) => {
+    .then((json) => {
```
Instead of "magically" checking the dependencies, what if we simply do `['dependencies.standard', 'devDependencies.standard']` for standardjs?

This means that this function becomes simpler and there's less magic.

chrisdothtml replied:
I like that approach. I can add these updates when I'm home from work

satazor commented:
@chrisdothtml I'm really sorry for the delay. I've made the review.

chrisdothtml committed:
adding support for object path lookup in tryPackageJson

chrisdothtml committed:
adding missing checks in linters

chrisdothtml committed:
adding support for xo

chrisdothtml commented:
@satazor added the updates, went through every linter and added dependency check (after looking up whether it was necessary), and also added xo support to close #13 for good

chrisdothtml commented:
@satazor please take a look at my updates and let me know if there's anything else required to get this merged

satazor commented on lib/util/tryPackageJson.js:
```diff
@@ -3,9 +3,36 @@
 const loadJsonFile = require('load-json-file');
 
-function tryPackageJson(dir, prop) {
+function objectPath(obj, propPath) {
```
Can you instead use ’lodash.get’ package instead of rolling out our own function?

chrisdothtml replied:
done

chrisdothtml replied:
anything else?

chrisdothtml committed:
replacing objectPath function with lodash.get

satazor commented:
Thank you and sorry for the time it took. I will release it as a feature.

chrisdothtml commented:
No worries. Thanks 👍
